### ISSUE TITLE
Cached Paths

### ISSUE THREAD

**Issue by Spedwards** [ScreepsOCS/screeps.behaviour-action-pattern#869]
So I've had this thought for cached paths.

```
Memory.rooms['<roomName>'].paths = {
    someDestID: {
        route: [
            {x: 10, y: 20, roomName: 'W0N0', nextDirection: 2},
            {x: 11, y: 19, roomName: 'W0N0', nextDirection: 8},
        ],
    }
}
```

You would have a set list of paths. The object key would be either the ID for a structure, or a flag name. The route would contain a list of room positions with the direction required for the creep to get to the next position.

The paths should be set from a spawn. If a creep starts them from any other position, they should look for the closest position in the route array, move there, and start their path.

The creep's data property should contain the index for the route the creep is up to, and the ID for the route. So using the above format, you could get it down to a one-liner once the creep is on the path:

`creep.move(creep.room.paths[creep.data.pathID].route[creep.data.pathIndex++].nextDirection);`

**Comment by Spedwards** [comment]
By storing the room positions, you are able to also efficiently go through rooms. The only problem with the system laid out, is when `creep.room` does not equal the room the path belongs to.

**Comment by chobobobo** [comment]
I like this, my current thinking is to break down the overall route even further, into a list of rooms that need to be traversed, then as the creep is going room by room it could load paths from a cache like this to get the directions to get it from the entrance to the exit within that room.

**Comment by Spedwards** [comment]
Another thought which would solve the problem aforementioned. Change the structure to:

```
Memory.paths = {
    someDestID: {
        startRoomA: {
            route: [
                {x: 10, y: 20, roomName: 'W0N0', nextDirection: 2},
                {x: 11, y: 19, roomName: 'W0N0', nextDirection: 8},
            ],
        },
        startRoomB: {
            // ...
        }
    }
}
```

Then all you would have to do is additionally save the starting room to the creep's data. Allows a path to have multiple starting points. Since paths can go through rooms and different rooms can have paths to the same dest, it seems better this way.

**Comment by chobobobo** [comment]
Would this allow two paths ABC and ABCD to share the same route information for rooms BC without duplication?

**Comment by logxen** [comment]
they should be able to share B ... C maybe not though

**Comment by Spedwards** [comment]
It could. You could have separate smaller routes from room to room, which the more specific routes could use if need be.

**Comment by chobobobo** [comment]
Yeah, I think that separate smaller routes from room to room is what I would like to try first, the route will be different based on what the next destination will be (choose the best exit).  Also, it's possible to use pathfinding to find the closest room position to join the route which should not be that expensive, so joining the route mid-stream or pathing around an obstacle should not be that expensive.

**Comment by chobobobo** [comment]
```Memory.rooms['<roomName>'].paths = {
    someDestID: {
        route: [
            {x: 10, y: 20, roomName: 'W0N0', nextDirection: 2},
            {x: 11, y: 19, roomName: 'W0N0', nextDirection: 8},
        ],
    }
}```

**Comment by chobobobo** [comment]
So for remoteMining, what I propose is that in Task.mining we store the route in terms of room traversal with the destination position in that room, where A,B,C,D are room names, and A, E are names of spawnRooms
```
Task.mining['C'].routes = {
	spawnRoomA: [{A, exitBCPos}, {B, exitCPos}, {C, sourcePos}]
	spawnRoomB: ...
}
```
in each room we would have paths between locations within that room including exits
```
Memory.rooms['A'].paths = {
	storageId-exitBC: {
		route: [
			{x:10, y:20, nextDirection:2},
			{x:11, y:19, nextDirection:8},
		]
	}
}
```
The reasoning here is that for each remote room we do one expensive routing call (which creeps do each time they go now) and then let the rooms determine the optimal paths between exits locally.

The challenging part remains determining which exit to take from a room for your destination, but that could be determined by looking at the exits in the initial route.  The only thing missing would be detection of needing to use new exits because one is blocked, so intelligent decisions about when to invalidate the route to the remote mining room.

**Comment by chobobobo** [comment]
I wonder though if the savings of reusing routes within rooms is really worth the extra effort or whether it might just be simpler to use the data structure you described initially, and accept that two routes to different destinations in the same room might overlap in intermediate rooms.. if we're storing all of these routes in globals instead of memory is it really worth the effort to reduce their size?

**Comment by chobobobo** [comment]
@Spedwards I think to avoid premature optimization we should go ahead and implement your first proposal and see how it runs, then consider whether we should optimize further by breaking up routes.

**Comment by Spedwards** [comment]
Being worked on in #1120.
